File: Java/601-650/AddOneRowToTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class AddOneRowToTree {
    public TreeNode addOneRow(TreeNode root, int v, int d) {
        if (d == 1) {
            TreeNode node = new TreeNode(v);
            node.left = root;
            return node;
        }
        traverse(root, v, d, 1);
        return root;
    }
    void traverse(TreeNode root, int v, int d, int depth) {
        if (root == null) return;
        if (depth == d - 1) {
            TreeNode left = new TreeNode(v);
            left.left = root.left;
            root.left = left;
            TreeNode right = new TreeNode(v);
            right.right = root.right;
            root.right = right;
            return;
        }
        traverse(root.left, v, d, depth + 1);
        traverse(root.right, v, d, depth + 1);
    }
}

File: Python/601-650/AddOneRowToTree.py
class Solution:
    def addOneRow(self, root, v, d):
        """
        :type root: TreeNode
        :type v: int
        :type d: int
        :rtype: TreeNode
        """
        if d == 1:
            n = TreeNode(v)
            n.left = root
            return n
        self.addOneRow1(root, v, d, 1)
        return root
    def addOneRow1(self, root, v, d, curr):
        if root == None:
            return
        if curr == d - 1:
            l, r = TreeNode(v), TreeNode(v)
            l.left, r.right = root.left, root.right
            root.left, root.right = l, r
            return
        self.addOneRow1(root.left, v, d, curr + 1)
        self.addOneRow1(root.right, v, d, curr + 1)
